src/main/scala/firrtl/passes/AnnotateMemMacros.scala

```diff
@@ -92,55 +102,45 @@ object AnalysisUtils {
-      val connects = getConnects(m)
-
-      // returns # of mask bits if used
-      def getMaskBits(wen: String, wmask: String): Option[Int] = {
-        val wenOrigin = we(getConnectOrigin(connects, wen))
-        val one1 = we(one)
-        val wmaskOrigin = connects.keys.toSeq.filter(_.startsWith(wmask)).map(x => we(getConnectOrigin(connects, x)))
-        // all wmask bits are equal to wmode/wen or all wmask bits = 1(for redundancy checking)
-        val redundantMask = wmaskOrigin.map( x => (x == wenOrigin) || (x == one1) ).foldLeft(true)(_ && _)
-        if (redundantMask) None else Some(wmaskOrigin.length)
-      }
-
-      def updateStmts(s: Statement): Statement = s match {
-        // only annotate memories that are candidates for memory macro replacements
-        // i.e. rw, w + r (read, write 1 cycle delay)
-        case m: DefMemory if m.readLatency == 1 && m.writeLatency == 1 && 
-            (m.writers.length + m.readwriters.length) == 1 && m.readers.length <= 1 =>
-          val dataBits = bitWidth(m.dataType)
-          val rwMasks = m.readwriters map (w => getMaskBits(s"${m.name}.$w.wmode", s"${m.name}.$w.wmask"))
-          val wMasks = m.writers map (w => getMaskBits(s"${m.name}.$w.en", s"${m.name}.$w.mask"))
-          val maskBits = (rwMasks ++ wMasks).head
-          val memAnnotations = Map("useMacro" -> true)
-          val tempInfo = appendInfo(m.info, memAnnotations)
-          if (maskBits == None) m.copy(info = tempInfo)
-          else m.copy(info = tempInfo.append("maskGran" -> dataBits/maskBits.get))       
-        case b: Block => b map updateStmts
-        case s => s
+  def updateStmts(connects: Connects)(s: Statement): Statement = s match {
+    // only annotate memories that are candidates for memory macro replacements
+    // i.e. rw, w + r (read, write 1 cycle delay)
+    case m: DefMemory if m.readLatency == 1 && m.writeLatency == 1 &&
+        (m.writers.length + m.readwriters.length) == 1 && m.readers.length <= 1 =>
+      val dataBits = bitWidth(m.dataType)
+      val rwMasks = m.readwriters map (rw =>
+        getMaskBits(connects, memPortField(m, rw, "wmode"), memPortField(m, rw, "wmask")))
+      val wMasks = m.writers map (w =>
+        getMaskBits(connects, memPortField(m, w, "en"), memPortField(m, w, "mask")))
+      val memAnnotations = Map("useMacro" -> true)
+      val tempInfo = appendInfo(m.info, memAnnotations)
+      (rwMasks ++ wMasks).head match {
+        case None =>
+          m copy (info = tempInfo)
+        case Some(maskBits) =>
+          m.copy(info = tempInfo.append("maskGran" -> dataBits / maskBits))
       }
-      m.copy(body=updateStmts(m.body))
-    } 
-
-    val updatedMods = c.modules map {
-      case m: Module => annotateModMems(m)
-      case m: ExtModule => m
-    }
-    c.copy(modules = updatedMods)
+    case s => s map updateStmts(connects)
+  }
 
-  }  
+  def annotateModMems(m: DefModule) = m map updateStmts(getConnects(m))
```

Comment:
ok but the return type is a DefModule?
which is either the original ExtModule or "Module" with whatever changes?

Haha... I'm still such a functional programming noob